dataset/src/main/scala/frameless/ops/SmartProject.scala

```diff
@@ -0,0 +1,58 @@
+package frameless
+package ops
+
+import shapeless.ops.hlist.ToTraversable
+import shapeless.ops.record.{Keys, SelectAll, Values}
+import shapeless.{HList, LUBConstraint, LabelledGeneric}
+
+import scala.annotation.implicitNotFound
+
+@implicitNotFound(msg = "Cannot prove that ${T} can be projected to ${U}. Perhaps not all member names and types of ${U} are the same in ${T}?")
+case class SmartProject[T: TypedEncoder, U: TypedEncoder](apply: TypedDataset[T] => TypedDataset[U])
+
+object SmartProject {
+  /**
+    * Proofs that there is a type safe projection from a type T to another type U. It requires that:
+    * (a) both T and U and products for which a LabelledGeneric can be derived (e.g., case classes),
+    * (b) all members of U have a corresponding member in T that has the same name and type.
+    *
+    * @param tgen the LabelledGeneric derived for T
+    * @param ugen the LabelledGeneric derived for U
+    * @param keys the keys of U
+    * @param lub confirms that the keys of U are all Symbols
+    * @param select selects all the keys of U from T
+    * @param values selects all the values of LabeledGeneric[U]
+    * @param typeEqualityProof proof that U and the projection of T have the same type
+    * @param keysTraverse allows for traversing the keys of U
+    * @tparam T the original type T
+    * @tparam U the projected type U
+    * @tparam TRec shapeless' Record representation of T
+    * @tparam TProj the projection of T using the keys of U
+    * @tparam URec shapeless' Record representation of U
+    * @tparam UVals the values of U as an HList
+    * @tparam UKeys the keys of U as an HList
+    * @return
+    */
+  implicit def deriveProduct[
+  T: TypedEncoder,
+  U: TypedEncoder,
+  TRec <: HList,
+  TProj <: HList,
+  URec <: HList,
+  UVals <: HList,
+  UKeys <: HList](
+    implicit
+    tgen: LabelledGeneric.Aux[T, TRec],
+    ugen: LabelledGeneric.Aux[U, URec],
+    keys: Keys.Aux[URec, UKeys],
+    lub: LUBConstraint[UKeys, Symbol],
+    select: SelectAll.Aux[TRec,UKeys,TProj],
```

Comment:
I think `TProj` is the same type than `UKeys`. If so, it means code could be simplified with:

- replacing all instances of `TProj` by `UKeys` (and removing the `TProj` type argument)
- Changing this `select: SelectAll.Aux[TRec,UKeys,TProj]` to `select: SelectAll[TRec,UKeys],`